### ISSUE TITLE
FR: Native Android Keyboard Support

### ISSUE THREAD

**Issue by edent** [koreader/koreader#7423]
**Does your feature request involve difficulty completing a task? Please describe.**

I find typing on KOReader's keyboard is slow and error prone.

* It doesn't support easy capitalisation https://github.com/koreader/koreader/issues/5975
* It requires a new build for every language https://github.com/koreader/koreader/issues/5615
* Doesn't work as well as native keyboards for some languages https://github.com/koreader/koreader/issues/6402
* Punctuation is in a different place to usual Android keyboards
* Doesn't support swipe to type
* It is slow to respond https://github.com/koreader/koreader/issues/6328
* Can't easily switch between languages

**Describe the solution you'd like**
I would like to use my Android device's native keyboard. 

I think many of the above problems could be fixed if there was the option to use the native keyboard.

When I use the Linux version, my desktop's physical keyboard works.

I understand that not every device has a native keyboard. I don't propose removing the current keyboard, but making it optional.

**Describe alternatives you've considered**
I've been using the KOReader keyboard - but it is really slow compared to my eReader's native keyboard.  When I use a different eReading app, it's much easier to take notes.  I vastly prefer KOReader's reading experience, but I need to take lots of notes - and it is hard.

**Additional context**
I love KOReader - but I wish it had better keyboard integration.

I notice a few years ago this sort of suggestion was marked as wontfix https://github.com/koreader/koreader/issues/2241#issuecomment-699682129 - has that changed?

**Comment by NiLuJe** [comment]
Probably not. A physical keyboard is doable, an Android IM, I think not?

**Comment by edent** [comment]
Just to be clear, I don't mean bundling a new IME. 

I mean letting the Android device's keyboard work in the same way as the desktop version's.

Does that make a difference?

**Comment by NiLuJe** [comment]
I didn't mean building a new IME, I meant integrating with native IME (which, err, isn't a thing on desktop either? the emu uses VirtualKeyboard like everybody else, it just happens to have basic physical keyboard support).

**Comment by poire-z** [comment]
We can use our physical keyboard on the emulator since #7037. Still need #7183 to have more keys working.

About Android, we're not a native Android application, so everything of the kind requested is complicated.

**Comment by edent** [comment]
I see. In a native Android app, I'd add to a layout:

```
android:inputType="textMultiLine|textLongMessage|textCapSentences"
```

Is there a way to do that in Lua? Pop up a native keyboard?  (I'm not familiar with Lua.)

**Comment by Frenzie** [comment]
Yes, except Android is leaps and bounds more complex to use than SDL.

**Comment by pazos** [comment]
> I notice a few years ago this sort of suggestion was marked as wontfix #2241 (comment) - has that changed?

Nope AFAICT. Won't fix means nobody is going to fix it but patches very welcome. I lurked a bit a couple of years ago. It is a major PITA, with some difficult parts:

1. You can invoke your preferred IME using JNI (pure native methods are still broken after more than 10 years :/), that's not a problem. Both show/hide work fine.
2. You need to resize the contents of the native window. That's the first problem. The android view hierarchy handles that in a way not available to use. We're drawing in a buffer, we support buffer resizes but we don't expect that part of the buffer is handled by an external application.
3. You need key translation functions. Each time the user presses a key we receive a bunch of key sequence events (onNativeKeyDown/Up) that we need to translate to the chars displayed on the screen.

The 3rd point is doable and somehow required get bluetooth/physical keyboards support. **But** the `keyboard is shown, get its size, adapt the interface to the buffer - IME size` part is wait to much work FWIW.

I also tried to get inspiration in how it is handled by other Native Activities. All of them are games or game-alike (like Scummvm or emulators) and all of them use a separate activity hooked in the view hierarchy to handle key inputs. Even more: all of them switched to NativeActivity to some sort of Activity and SurfaceView, implementing just the render in native code, and using native functions to tell the engine where to start/stop [*]

We're the only native activity out there that's both 1. opensource 2. not a game 3. implemented entirely using the ANativeActivity framework. That's fun but has some drawbacks. This is one of them. Sadly switching to Activity + SurfaceView is not an option for us as we need/want LuaJIT running the app main loop and that's not possible using managed activities.

*. This is a common approach in a bunch of apps. For instance: Coolreader for android is implemented on top of an android activity plus a few native methods to invoke crengine.

**Comment by beelzebielsk** [comment]
@pazos Could you elaborate on this point? I'd like support for a bluetooth keyboard on android. Let me know if I should open a new issue for bluetooth keyboard support to talk about this.

> 3. You need key translation functions. Each time the user presses a key we receive a bunch of key sequence events (onNativeKeyDown/Up) that we need to translate to the chars displayed on the screen.
> 
> The 3rd point is doable and somehow required to get bluetooth/physical keyboards support.

**Comment by pazos** [comment]
@beelzebielsk: hi, opened a new ticket in android-luajit-launcher. 99% of development needs to happen there. Integration with koreader-base/koreader is the easiest part.

> I'd like support for a bluetooth keyboard on android.

I would love to see that too. You don't need to have experience with lua or koreader, but a basic knowledge of kotlin and android activities is required. See https://github.com/koreader/android-luajit-launcher/issues/338#issuecomment-961005237

**Comment by arooni** [comment]
As I read increasingly on my S8 Ultra tablet during the day, and my Boox tablet in the evening, both Android tablets of different flavors... I would very much like to see this added to koreader.  It would unlock the ability to take notes on both devices with a couple of bluetooth hardware keyboards I have around the house. 

Sometimes I have some insights that I feel I just don't want to finger type... and perhaps the solution is to switch apps to something like vim where I can take notes and come back; but I love the annotation idea and quoting ideas/concepts in context that KoReader allows me.

**Comment by arooni** [comment]
> Nope AFAICT. Won't fix means nobody is going to fix it but patches very welcome. I lurked a bit a couple of years ago. It is a major PITA, with some difficult parts:
> 
> 1. You can invoke your preferred IME using JNI (pure native methods are still broken after more than 10 years :/), that's not a problem. Both show/hide work fine.
> 2. You need to resize the contents of the native window. That's the first problem. The android view hierarchy handles that in a way not available to use. We're drawing in a buffer, we support buffer resizes but we don't expect that part of the buffer is handled by an external application.
> 3. You need key translation functions. Each time the user presses a key we receive a bunch of key sequence events (onNativeKeyDown/Up) that we need to translate to the chars displayed on the screen.
> 
> The 3rd point is doable and somehow required to get bluetooth/physical keyboards support. **But** the `keyboard is shown, get its size, adapt the interface to the buffer - IME size` part is way to much work FWIW.
> 

Dumb question, but could you just invoke the keyboard, not redraw the screen so just let it overlap?  I guess you'd still need to respond to each onKeyDown, OnKeyUP in a loop.

**Comment by charleshan** [comment]
It's really unfortunate to hear that native android keyboard support is unlikely. I can type 4x faster with the native android keyboard with the familiar keyboard layout (muscle memory) and auto-correct

**Comment by bassmannate** [comment]
I agree it would be nice but I get that this is all geared toward e-ink devices. Heck, I'd be happy if I had the option to hide the on screen keyboard in the desktop/SDL port since I have a physical keyboard on those devices.

**Comment by Frenzie** [comment]
You can by setting `isTouchDevice = no` but that might potentially have some unintended side effects.

**Comment by charleshan** [comment]
What if we created a basic Android app to take advantage of "display over other apps" to type notes using the native keyboard? We can open the external app using a deep link and maybe pass the text into KOReader using the clipboard.

**Comment by charleshan** [comment]
I found a workaround to use native android keyboard using the Automator app. I just need KOReader to copy to clipboard automatically when "Edit Note" is open. What am I doing wrong here? KOReader crashes without any explanation whenever `Edit Note` opens.

```lua
function ReaderBookmark:setBookmarkNote(item_or_index, is_new_note, new_note, caller_callback)
    local item, index
    if self.bookmark_menu then
        item = item_or_index -- in item_table
        index = self:getBookmarkItemIndex(item)
    else -- from Highlight
        index = item_or_index
    end
    local annotation = self.ui.annotation.annotations[index]
    local type_before = item and item.type or self.getBookmarkType(annotation)
    local input_text = annotation.note
    if new_note then
        if input_text then
            input_text = input_text .. "\n\n" .. new_note
        else
            input_text = new_note
        end
    end
    local input_dialog
    input_dialog = InputDialog:new{
        title = _("Edit note"),
        description = "   " .. self:_getDialogHeader(annotation),
        input = input_text,
        allow_newline = true,
        add_scroll_buttons = true,
        use_available_height = true,
        buttons = {
            {
                {
                    text = _("Cancel"),
                    id = "close",
                    callback = function()
                        UIManager:close(input_dialog)
                        if is_new_note then -- "Add note" called from highlight dialog and cancelled, remove saved highlight
                            self:removeItemByIndex(index)
                        end
                    end,
                },
                {
                    text = _("Paste"), -- insert highlighted text
                    callback = function()
                        input_dialog:addTextToInput(annotation.text)
                    end,
                },
                {
                    text = _("Save"),
                    is_enter_default = true,
                    callback = function()
                        local value = input_dialog:getInputText()
                        if value == "" then -- blank input deletes note
                            value = nil
                        end
                        annotation.note = value
                        local type_after = self.getBookmarkType(annotation)
                        if type_before ~= type_after then
                            if type_before == "highlight" then
                                self.ui:handleEvent(Event:new("AnnotationsModified",
                                    { annotation, nb_highlights_added = -1, nb_notes_added = 1 }))
                            else
                                self.ui:handleEvent(Event:new("AnnotationsModified",
                                    { annotation, nb_highlights_added = 1, nb_notes_added = -1 }))
                            end
                        end
                        UIManager:close(input_dialog)
                        if item then
                            item.note = value
                            item.type = type_after
                            item.text = self:getBookmarkItemText(item)
                        end
                        caller_callback()
                    end,
                },
            }
        },
    }
    UIManager:show(input_dialog)
    input_dialog:onShowKeyboard()
    Device.input.setClipboardText(input_text) -- ADDED
end
```

**Comment by hius07** [comment]
> ```lua
> Device.input.setClipboardText(input_text) -- ADDED
> ```

I do not see anything wrong here, the crash is not related to this line.

**Comment by charleshan** [comment]
I think it was. I add a null check and that fixed the issue.

```lua
-- Ensure input_text is not nil or empty
    if not input_text or input_text == "" then
        input_text = ""
    end
    -- Set clipboard text
    local success, err = pcall(function()
        Device.input.setClipboardText(input_text)
    end)
    if not success then
        print("Error setting clipboard text: " .. tostring(err))
    end
```

**Comment by hius07** [comment]
Maybe device specific. Doesn't crash on Kindle when setting the clipboard to `nil`.

So you can have
```Lua
Device.input.setClipboardText(input_text or "")
```

**Comment by charleshan** [comment]
Thanks! That's a lot cleaner.

I got it to work! Automator watches the clipboard and whenever `Edit Note` is triggered, an input modal from Automator is displayed with the native android keyboard. When I press `ok`, it runs a series of automations to copy text input to clipboard, paste to KOReader `Edit Note` input, and taps save. Likewise with `cancel`.

4x typing speed feels good!

**Comment by awsharif** [comment]
> Thanks! That's a lot cleaner.
> 
> I got it to work! Automator watches the clipboard and whenever `Edit Note` is triggered, an input modal from Automator is displayed with the native android keyboard. When I press `ok`, it runs a series of automations to copy text input to clipboard, paste to KOReader `Edit Note` input, and taps save. Likewise with `cancel`.
> 
> The typing experience has been the one thing I didn't like about KOReader and that's now been solved :)
> 
> Here's a quick demo. It was recorded on Boox Poke 3 (e-ink). The video is at 1x. Voice-to-text works as well so I can type even faster.
> screen-20240805-222609.mp4

This is so cool, could you share how you implement this, please?

**Comment by jm355** [comment]
This seems worth initiating a pull request for, and then let it be opitimized in the future. The sooner we can start using this the better :)

**Comment by charleshan** [comment]
I already shared how to implement this in my comment above. 

> Automate (android app) watches the clipboard and whenever Edit Note is triggered, an input modal from Automator is displayed with the native android keyboard. When I press ok, it runs a series of automations to copy text input to clipboard, paste to KOReader Edit Note input, and taps save. Likewise with cancel.

You need two things:
1. Build a custom apk with the changes above
2. Use a third party app to watch the clipboard and automate a set of tasks (explained above)

For step 2, the automation is device specific because you need to set the coordinates for each tap automation.

I hope this helps!

![image](https://github.com/user-attachments/assets/4dfe83a5-7096-4781-a10c-b41cb56d3166)

**Comment by hius07** [comment]
If this approach is popular, we can think about adding this (or similar) line to the InputText module, to fire automation on any input dialog appearance.

**Comment by charleshan** [comment]
It's a possibility. Currently, my automation also taps `Save` or `Cancel` for editing notes so I would have to pass a value through the clipboard to differentiate between other input dialogs.

**Comment by uroybd** [comment]
> If this approach is popular, we can think about adding this (or similar) line to the InputText module, to fire automation on any input dialog appearance.

1. It can be an optional thing set in keyboard settings.
2. Or, we can provide a button beside text inputs to launch the external input editor. For a minuscule amount of text, a built-in keyboard is just fine.

**Comment by Bluemoondragon07** [comment]
> I found a workaround to use native android keyboard using the Automate app. I just need KOReader to copy to clipboard automatically when "Edit Note" is open. What am I doing wrong here? KOReader crashes without any explanation whenever `Edit Note` opens.
> 
> ```lua
> function ReaderBookmark:setBookmarkNote(item_or_index, is_new_note, new_note, caller_callback)
>     local item, index
>     if self.bookmark_menu then
>         item = item_or_index -- in item_table
>         index = self:getBookmarkItemIndex(item)
>     else -- from Highlight
>         index = item_or_index
>     end
>     local annotation = self.ui.annotation.annotations[index]
>     local type_before = item and item.type or self.getBookmarkType(annotation)
>     local input_text = annotation.note
>     if new_note then
>         if input_text then
>             input_text = input_text .. "\n\n" .. new_note
>         else
>             input_text = new_note
>         end
>     end
>     local input_dialog
>     input_dialog = InputDialog:new{
>         title = _("Edit note"),
>         description = "   " .. self:_getDialogHeader(annotation),
>         input = input_text,
>         allow_newline = true,
>         add_scroll_buttons = true,
>         use_available_height = true,
>         buttons = {
>             {
>                 {
>                     text = _("Cancel"),
>                     id = "close",
>                     callback = function()
>                         UIManager:close(input_dialog)
>                         if is_new_note then -- "Add note" called from highlight dialog and cancelled, remove saved highlight
>                             self:removeItemByIndex(index)
>                         end
>                     end,
>                 },
>                 {
>                     text = _("Paste"), -- insert highlighted text
>                     callback = function()
>                         input_dialog:addTextToInput(annotation.text)
>                     end,
>                 },
>                 {
>                     text = _("Save"),
>                     is_enter_default = true,
>                     callback = function()
>                         local value = input_dialog:getInputText()
>                         if value == "" then -- blank input deletes note
>                             value = nil
>                         end
>                         annotation.note = value
>                         local type_after = self.getBookmarkType(annotation)
>                         if type_before ~= type_after then
>                             if type_before == "highlight" then
>                                 self.ui:handleEvent(Event:new("AnnotationsModified",
>                                     { annotation, nb_highlights_added = -1, nb_notes_added = 1 }))
>                             else
>                                 self.ui:handleEvent(Event:new("AnnotationsModified",
>                                     { annotation, nb_highlights_added = 1, nb_notes_added = -1 }))
>                             end
>                         end
>                         UIManager:close(input_dialog)
>                         if item then
>                             item.note = value
>                             item.type = type_after
>                             item.text = self:getBookmarkItemText(item)
>                         end
>                         caller_callback()
>                     end,
>                 },
>             }
>         },
>     }
>     UIManager:show(input_dialog)
>     input_dialog:onShowKeyboard()
>     Device.input.setClipboardText(input_text) -- ADDED
> end
> ```

@charleshan, is this a user patch? I've been trying to replicate your keyboard setup and understood this as a user patch to trigger clipboard changes for Automate to listen for. But every time I start KOreader with this patch, it says: "Error applying patch: <patch-name>".

**Comment by charleshan** [comment]
I created a new build with the modifications. I was not aware of user patches but I think that would be the easier and better approach. Did you import the required files? I'm not too familiar with lua but here's my guess at what that would look like to override `setBookmarkNote`.

```lua
local Device = require("device")
local ReaderBookmark = require("apps/reader/modules/readerbookmark")
local InputDialog = require("ui/widget/inputdialog")
local UIManager = require("ui/uimanager")
-- (might be missing some required imports)

local setBookmarkNote_orig = ReaderBookmark.setBookmarkNote
ReaderBookmark.setBookmarkNote = function setBookmarkNote(item_or_index, is_new_note, new_note, caller_callback)
    local item, index
    if self.bookmark_menu then
    -- (rest of the code)
```
***
Questions to devs: Is it best practice to initialize `setBookmarkNote_orig` when we're overriding `ReaderBookmark.setBookmarkNote`? I see that the `Notable patches` all do this and it seems unnecessary. 
https://github.com/koreader/koreader/wiki/User-patches

**Comment by poire-z** [comment]
> Is it best practice to initialize setBookmarkNote_orig when we're overriding ReaderBookmark.setBookmarkNote? 

The patches that do this are because in the function we replace, we are calling the original (instead of copying it all and miss future modifications), we just do our added stuff before or/and after calling it.
If you just want to replace it all, no need to keep the original around.
(The name with _orig is just my habit.)

**Comment by charleshan** [comment]
Ohhh I miss that bit. I was reading the gist in plaintext and I guess my brain got lazy lol

Thanks for the explanation :)

**Comment by pazos** [comment]
> It's really unfortunate to hear that native android keyboard support is unlikely. I can type 4x faster with the native android keyboard with the familiar keyboard layout (muscle memory) and auto-correct

@charleshan: It is unlikely to be contributed by **me**. Patches are very welcome  :)
I would gladly help on proper support once somebody contributes the time for implementing and testing the bits suggested on this ticket and linked ones.

I will update OP because some important things that were a PITA don't apply anymore.

To recap: for android keyboard support we need (in this order)

1. bluetooth/usb support (aka event tranlation)
2. write stub code to show and hide the keyboard in the android module
3. write stub code to switch between our virtualKeyboard and external ones

**Comment by isdamir** [comment]
> Thanks! That's a lot cleaner.
> 
> I got it to work! Automate (android app) watches the clipboard and whenever `Edit Note` is triggered, an input modal from Automator is displayed with the native android keyboard. When I press `ok`, it runs a series of automations to copy text input to clipboard, paste to KOReader `Edit Note` input, and taps save. Likewise with `cancel`.
> 
> The typing experience has been the one thing I didn't like about KOReader and that's now been solved :)
> 
> Here's a quick demo. It was recorded on Boox Poke 3 (e-ink). The video is at 1x. Voice-to-text works as well so I can type even faster.
> 
>  screen-20240805-222609.mp4

that's very cool ,i can try it.

**Comment by isdamir** [comment]
Although I wrote the userpatch, my device does not allow "display on top of other apps"

```
local Device = require("device")
local ReaderBookmark = require("apps/reader/modules/readerbookmark")

local setBookmarkNote_orig = ReaderBookmark.setBookmarkNote
ReaderBookmark.setBookmarkNote = function(self, item_or_index, is_new_note, new_note, caller_callback)
    setBookmarkNote_orig(self, item_or_index, is_new_note, new_note, caller_callback)
    local item, index
    if self.bookmark_menu then
        item = item_or_index -- in item_table
        index = self:getBookmarkItemIndex(item)
    else -- from Highlight
        index = item_or_index
    end
    local annotation = self.ui.annotation.annotations[index]
    local input_text = annotation.note
    if new_note then
        if input_text then
            input_text = input_text .. "\n\n" .. new_note
        else
            input_text = new_note
        end
    end
    Device.input.setClipboardText(input_text or "")
end
```

**Comment by uroybd** [comment]
> I already shared how to implement this in my comment above.
> 
> > Automate (android app) watches the clipboard and whenever Edit Note is triggered, an input modal from Automator is displayed with the native android keyboard. When I press ok, it runs a series of automations to copy text input to clipboard, paste to KOReader Edit Note input, and taps save. Likewise with cancel.
> 
> You need two things:
> 
>     1. Build a custom apk with the changes above. This is to replace two steps of holding gesture of the text input to get the text actions menu to pop up and copy the text.
> 
>     2. Use a third party app to watch the clipboard and automate a set of tasks (explained above). I used a free app called Automate.
> 
> 
> For step 2, the automation is device specific because you need to set the coordinates for each tap automation.
> 
> I hope this helps!
> 
> ![image](https://github.com/user-attachments/assets/4dfe83a5-7096-4781-a10c-b41cb56d3166)

Can I get your `.flo` file please?

**Comment by Yalums** [comment]
Hello, I'm just checking the latest status of this feature request. Is there any progress or workaround for using system input methods with KOReader? Thanks!

**Comment by Yalums** [comment]
Testing

**Comment by Frenzie** [comment]
Some implementation notes here: https://github.com/koreader/koreader/discussions/14358#discussioncomment-14500895

There are two main issues:

1. I don't find Android fun.
2. I don't like the way I hacked together the Looper thing to send through the relevant events, but I don't know how to do it any better. As stated there, judging by SDL code maybe it's just the way it has to be done.

**Comment by charleshan** [comment]
@uroybd I shared a video in a previous comment https://github.com/koreader/koreader/issues/7423#issuecomment-2270391205 showing how much faster it is. But I guess the device could be a limiting factor.

> The typing experience has been the one thing I didn't like about KOReader and that's now been solved :)
> 
> Here's a quick demo. It was recorded on Boox Poke 3 (e-ink). The video is at 1x. Voice-to-text works as well so I can type even faster.
> 
>  [screen-20240805-222609.mp4](https://github.com/user-attachments/assets/72a8d7e6-ea4c-4f2a-9787-b6c6534f8b5c)

Of course, native support would be much better. I'm just patiently waiting for @Frenzie to make some breakthrough :)

**Comment by NikoKS** [comment]
hello, is this still difficult to implement?

**Comment by Frenzie** [comment]
It's quite a bit more challenging than I thought in <https://github.com/koreader/koreader/issues/7423#issuecomment-3436007386>, but I did manage to get it working with a somewhat frightening amount of code.

If there's anything still wrong with it other than potential visual overlaps, let me know. The build is today's nightly + the latest luajit-launcher + Android IME patches.
https://github.com/Frenzie/android-luajit-launcher/releases/tag/fake-tag-keyboard-poc

**Comment by Bluemoondragon07** [comment]
> https://github.com/Frenzie/android-luajit-launcher/releases/tag/fake-tag-keyboard-poc

 @Frenzie Wow, this works great. No bugs yet so far. Weirdly, this drastically improved the refresh rate as I'm typing (Meebook M7). Typing with KOreader's keyboard is usually pretty laggy. Also, tested in the text file editor. This also allows me to type in KOreader with a physical keyboard on Android, whereas before I couldn't.

**Comment by ryanwwest** [comment]
Works really great for me too.

**Comment by Bluemoondragon07** [comment]
Okay, one bug, sometimes when I turn on my Meebook while a book is open in KOreader and immediately touch the screen, it brings  up the android keyboard. Only sometimes, I don't know how to reproduce this at will.

**Comment by Frenzie** [comment]
It pops up using the startTextInput function that was originally added for SDL. That doesn't happen much: https://github.com/search?q=repo%3Akoreader%2Fkoreader%20startTextInput&type=code

**Comment by charleshan** [comment]
Thanks @Frenzie!

A few things:
- Backspace doesn't work with my main keyboard Gboard which is most popular android keyboard and is made by Google. Out of curiosity, I tested with another keyboard called Android Keyboard (AOSP) that I never use and backspace seems to work.
- The keyboard overlaps with the UI when the menu is vertically long but that's a common issue with responsive mobile design. A simple workaround is to press the back button to hide the keyboard and tap on the text input again when you're ready to type.

**Comment by Frenzie** [comment]
> Backspace doesn't work with my main keyboard Gboard 

For real? That's depressing. Implementing backspace was quite challenging, but I wasn't exactly counting on different software keyboards doing backspace differently…

Edit: in short, the way the IME system works is something like

this is text [current word]

Anything withing [current word] can be completely replaced at any time.

That makes the state management quite complex and hard to follow, and backspace in particular is a bit of a wild card since it should still act on single characters in [current word] without removing it completely.

I'm not entirely sure if IMEs are so complex by necessity or if Google just overcomplicated it.

My hope is that this implementation concept will also work on Mac/Linux with no or minimal changes, but I haven't looked into that yet. Especially because I'm afraid that Mac/Linux IMEs will be significantly easier to work with. ;-D

**Comment by Bluemoondragon07** [comment]
Interesting, AOSP, Typewise and Swiftkey dont have any problems with backspace.

**Comment by charleshan** [comment]
> I'm not entirely sure if IMEs are so complex by necessity or if Google just overcomplicated it.

I started using Swiftkey (made by Microsoft) and it seems to work well! Backspace, autocorrect, and word suggestion work perfectly.

If anyone wants to use Swiftkey keyboard, here are some Swiftkey settings to improve the typing experience on an e-ink device:
Layout & keys > Key pop-ups > Disable
Layout & keys > Long-press symbols > Enable
Typing > Cursor control (Tap and hold the spacebar to move the cursor) > Enable

Do we need anything else to get this merged?

**Comment by leadermontanus** [comment]
Just installed it on Bigme B6 eink device, works perfectly with SwiftKey. The keyboard issue was the only thing that kept me from using Koreader for work reading (I add a lot of notes as I read), so now I can use Koreader for both leisure and work. 

Fantastic work!

**Comment by titandrive** [comment]
this issue has been driving me nuts as well as I can't stand the stock keyboard, especially on my anddroid phone. this thread inspired me to create a plugin that solves the same issue, while we wait for this to be merged. I think ive more or less solved all of the issues, namely the backspace issue, noted above. 

https://github.com/titandrive/koboard

**Comment by edent** [comment]
@titandrive this works perfectly for me. Thank you so much!

**Comment by gerardomtz26** [comment]
> this issue has been driving me nuts as well as I can't stand the stock keyboard, especially on my anddroid phone. this thread inspired me to create a plugin that solves the same issue, while we wait for this to be merged. I think ive more or less solved all of the issues, namely the backspace issue, noted above.
> 
> https://github.com/titandrive/koboard

I created a fork of the original KOBoard plugin that adds SwiftKey suggestions and basic autocorrection support:

⁠https://github.com/gerardomtz26/koboard

The fork extends the original plugin’s Android IME bridge to provide enough editable state for SwiftKey to show predictive suggestions and perform basic autocorrection during normal typing.

I tested it on my Bigme HiBreak Pro with KOReader, and it works well for my normal typing and note-taking use case.

There are still some limitations, particularly when manually moving the cursor within existing text, since the IME may not immediately receive complete contextual information.

I’m sharing it here in case it is useful to other Android KOReader users.

**Comment by titandrive** [comment]
> > this issue has been driving me nuts as well as I can't stand the stock keyboard, especially on my anddroid phone. this thread inspired me to create a plugin that solves the same issue, while we wait for this to be merged. I think ive more or less solved all of the issues, namely the backspace issue, noted above.
> > https://github.com/titandrive/koboard
> 
> I created a fork of the original KOBoard plugin that adds SwiftKey suggestions and basic autocorrection support:
> 
> ⁠https://github.com/gerardomtz26/koboard
> 
> The fork extends the original plugin’s Android IME bridge to provide enough editable state for SwiftKey to show predictive suggestions and perform basic autocorrection during normal typing.
> 
> I tested it on my Bigme HiBreak Pro with KOReader, and it works well for my normal typing and note-taking use case.
> 
> There are still some limitations, particularly when manually moving the cursor within existing text, since the IME may not immediately receive complete contextual information.
> 
> I’m sharing it here in case it is useful to other Android KOReader users.

You good if I merge these changes into mine?

**Comment by gerardomtz26** [comment]
> > > this issue has been driving me nuts as well as I can't stand the stock keyboard, especially on my anddroid phone. this thread inspired me to create a plugin that solves the same issue, while we wait for this to be merged. I think ive more or less solved all of the issues, namely the backspace issue, noted above.
> > > https://github.com/titandrive/koboard
> > 
> > 
> > I created a fork of the original KOBoard plugin that adds SwiftKey suggestions and basic autocorrection support:
> > ⁠https://github.com/gerardomtz26/koboard
> > The fork extends the original plugin’s Android IME bridge to provide enough editable state for SwiftKey to show predictive suggestions and perform basic autocorrection during normal typing.
> > I tested it on my Bigme HiBreak Pro with KOReader, and it works well for my normal typing and note-taking use case.
> > There are still some limitations, particularly when manually moving the cursor within existing text, since the IME may not immediately receive complete contextual information.
> > I’m sharing it here in case it is useful to other Android KOReader users.
> 
> You good if I merge these changes into mine?

Of course, merge it. No problem at all. 
Thanks for the original plugin.
